kubernetes-client/src/Kubernetes/Client/Config.hs

```diff
@@ -1,136 +1,158 @@
+  , parsePEMCerts
+  , serviceAccountDir
+  , setCAStore
+  , setClientCert
+  , setMasterURI
+  , setTokenAuth
+  , tlsValidation
+  )
+where
+
+import qualified Kubernetes.OpenAPI.Core as K
+
+import           Control.Applicative                 ((<|>))
+import           Control.Exception.Safe              (MonadThrow, throwM)
+import           Control.Monad.IO.Class              (MonadIO, liftIO)
+import qualified Data.ByteString                     as B
+import qualified Data.ByteString.Base64              as B64
+import qualified Data.ByteString.Lazy                as LazyB
+import           Data.Either.Combinators
+import           Data.Function                       ((&))
+import           Data.Maybe
+import qualified Data.Text                           as T
+import qualified Data.Text.Encoding                  as T
+import qualified Data.Text.IO                        as T
+import           Data.Yaml
+import           Kubernetes.Client.Auth.ClientCert
+import           Kubernetes.Client.Auth.GCP
+import           Kubernetes.Client.Auth.OIDC
+import           Kubernetes.Client.Auth.Token
+import           Kubernetes.Client.Internal.TLSUtils
+import           Kubernetes.Client.KubeConfig
+import           Network.Connection                  (TLSSettings (..))
+import qualified Network.HTTP.Client                 as NH
+import           Network.HTTP.Client.TLS             (mkManagerSettings)
+import qualified Network.TLS                         as TLS
+import           System.Environment                  (getEnv)
+import           System.FilePath
+
+data KubeConfigSource = KubeConfigFile FilePath
+                      | KubeConfigCluster
+
+{-|
+  Creates 'NH.Manager' and 'K.KubernetesClientConfig' for a given
+  'KubeConfigSource'. It is recommended that multiple 'kubeClient' invocations
+  across an application share an 'OIDCCache', this makes sure updation of OAuth
+  token is synchronized across all the different clients being used.
+-}
+kubeClient
+  :: OIDCCache
+  -> KubeConfigSource
+  -> IO (NH.Manager, K.KubernetesClientConfig)
+kubeClient oidcCache (KubeConfigFile f) = do
+  kubeConfigFile <- decodeFileThrow f
+  uri <- getCluster kubeConfigFile
+         & fmap server
+         & either (const $ pure "localhost:8080") return
+  t <- defaultTLSClientParams
+       & fmap (tlsValidation kubeConfigFile)
+       & (>>= (addCACertData kubeConfigFile))
+       & (>>= addCACertFile kubeConfigFile (takeDirectory f))
+  c <- K.newConfig & fmap (setMasterURI uri)
+  (tlsParams, cfg) <-
+    case getAuthInfo kubeConfigFile of
+      Left _          -> return (t,c)
+      Right (_, auth) -> applyAuthSettings oidcCache auth (t, c)
+  mgr <- newManager tlsParams
+  return (mgr, cfg)
+kubeClient _ (KubeConfigCluster) = Kubernetes.Client.Config.cluster
+
+-- |Creates 'NH.Manager' and 'K.KubernetesClientConfig' assuming it is being executed in a pod
+cluster :: (MonadIO m, MonadThrow m) => m (NH.Manager, K.KubernetesClientConfig)
```

Comment:
Same as before, `mkCluster`, `mkClusterClientConfig`, or some variant is probably more descriptive.

Reply:
Called it `mkInClusterClientConfig` as most other language clients also call it this.